### PR TITLE
add scan function

### PR DESCRIPTION
**Pull request by TimothyElems**
Close https://github.com/unifyai/ivy/issues/17449

### CODE REVIEW

**Comment by ivy-leaves** [comment]
If you are working on an open task, please **edit** the PR description to link to the issue you've created.

For more information, please check [ToDo List Issues Guide](https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues).

Thank you :hugs:

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@TimothyElems Thanks for making the PR! Please can you create an issue linked to scan of this issue https://github.com/unifyai/ivy/issues/1579? You can find more details about this here https://unify.ai/docs/ivy/overview/contributing/the_basics.html#todo-list-issues. 

Please could you also write a test for the frontend function you've implemented? You can find details about writing frontend tests here https://unify.ai/docs/ivy/overview/deep_dive/ivy_frontends_tests.html#ivy-frontend-tests.

**Comment by TimothyElems** [comment]
https://github.com/unifyai/ivy/issues/17449

**Comment by TimothyElems** [comment]
@Sam-Armstrong I used some previous tests from the same module and made the proper adjustments

[Scan function's test #17450](https://github.com/unifyai/ivy/pull/17450)

**Comment by Sam-Armstrong** [comment]
@TimothyElems please could you make the changes part of this pull request? You can do this by pushing to the same branch of your fork as before.

**Comment by TimothyElems** [comment]
@Sam-Armstrong 
I've been stuck on a check. All others work but one. And it seems the bug was not caused by anything I interacted with. What could the solution be?

File: ivy/functional/frontends/jax/lax/control_flow_operators.py [hunk 1]
```diff
@@ -58,3 +58,15 @@ def while_loop(cond_fun, body_fun, init_val):
     while cond_fun(val):
         val = body_fun(val)
     return val
+
+
+@to_ivy_arrays_and_back
+def scan(f, init, xs, length=None):
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/jax/lax/control_flow_operators.py:64]: This should take the same args as the original function, in this case (f, init, xs, length=None, reverse=False, unroll=1)

https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 2]
```diff
@@ -203,3 +204,50 @@ def _test_body_fn(x):
...
+        input_dtypes=input_dtype,
+        test_flags=test_flags,
+        frontend=frontend,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        f=_test_f,
+        # init=0,
+        xs=x,
+        length=length,
+        expected_result=expected_result,
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:252]: All function args need to be passed here as kwargs, including 'init' and the ones added in our modified function signature.

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 3]
```diff
@@ -203,3 +204,50 @@ def _test_body_fn(x):
...
+
+    input_dtype, x = dtype_and_x
+    length = len(x)
+    expected_carry = 0
+    expected_ys = []
+    carry = expected_carry
+    for elem in x:
+        carry, y = _test_f(carry, elem)
+        expected_ys.append(y)
+    expected_result = (expected_carry, ivy.stack(expected_ys))
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:240]: We don't need to define any expected result - this is all handled by helpers.test_frontend_function.

File: ivy/functional/frontends/jax/lax/control_flow_operators.py [hunk 4]
```diff
@@ -58,3 +58,15 @@ def while_loop(cond_fun, body_fun, init_val):
...
+    if xs is None:
+        xs = [None] * length
+    carry = init
+    ys = []
+    for x in xs:
+        carry, y = f(carry, x)
+        ys.append(y)
+    return carry, ivy.stack(ys)
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/jax/lax/control_flow_operators.py:72]: I'd recommend a good way of debugging this would be to create a python script which runs both the original jax.lax.scan function and your function for the same input. The return values for each function should be identical for all cases. But if you do this, do _not_ push the script to github.

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]

**Comment by TimothyElems** [comment]
@Sam-Armstrong 
The code works! My script works! The formats are clean! A new error persists?
Thoughts?

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 5]
```diff
@@ -203,3 +205,55 @@ def _test_body_fn(x):
...
+)
+def test_jax_lax_scan(
+    *,
+    dtype_and_x,
+    test_flags,
+    on_device,
+    fn_tree,
+    frontend,
+    unroll=1,
+):
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:228]:
```suggestion
    dtype_and_x,
    unroll,
    test_flags,
    on_device,
    fn_tree,
    frontend,
):
```

File: ivy/functional/frontends/jax/lax/control_flow_operators.py [hunk 6]
```diff
@@ -58,3 +58,21 @@ def while_loop(cond_fun, body_fun, init_val):
     while cond_fun(val):
         val = body_fun(val)
     return val
+
+
+@to_ivy_arrays_and_back
+def scan(f, init, xs, length=None, reverse=False, unroll=1):
+    if length is None:
+        length = len(xs[0]) if isinstance(xs[0], ivy.ndarray) else len(xs)
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/jax/lax/control_flow_operators.py:66]:
```suggestion
        length = len(xs[0]) if isinstance(xs[0], ivy.Array) else len(xs)
```

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@TimothyElems I've left a few comments of things that need changing. Can you send me an example of it working, so I can understand what is going on? Thanks

**Comment by TimothyElems** [comment]
@Sam-Armstrong 
ALL SET!! Passed all checks!

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

File: ivy/functional/frontends/jax/lax/control_flow_operators.py [hunk 7]
```diff
@@ -58,3 +58,21 @@ def while_loop(cond_fun, body_fun, init_val):
...
+
+    if reverse:
+        xs = xs[::-1]
+
+    ys = []
+    carry = init
+    for x in xs:
+        for _ in range(unroll):
+            carry = f(carry, x)
+            ys.append(carry)
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/jax/lax/control_flow_operators.py:76]:
```suggestion
            carry, y = f(carry, x)
            ys.append(y)
```

File: ivy/functional/frontends/jax/lax/control_flow_operators.py [hunk 8]
```diff
@@ -58,3 +58,21 @@ def while_loop(cond_fun, body_fun, init_val):
...
+
+    return ivy.stack(ys)
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/jax/lax/control_flow_operators.py:78]:
```suggestion
    return carry, ivy.stack(ys)
```

File: ivy/functional/frontends/jax/lax/control_flow_operators.py [hunk 9]
```diff
@@ -58,3 +58,21 @@ def while_loop(cond_fun, body_fun, init_val):
     while cond_fun(val):
         val = body_fun(val)
     return val
+
+
+@to_ivy_arrays_and_back
+def scan(f, init, xs, length=None, reverse=False, unroll=1):
```
Comment (Sam-Armstrong) [inline ivy/functional/frontends/jax/lax/control_flow_operators.py:64]: I think the way you are using 'unroll' and 'reverse' might be incorrect. Would recommend looking through the jax docs/source code to help solve this.
https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.scan.html
https://jax.readthedocs.io/en/latest/_modules/jax/_src/lax/control_flow/loops.html#scan

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 10]
```diff
@@ -203,3 +204,53 @@ def _test_body_fn(x):
...
+def test_jax_lax_scan(
+    *,
+    dtype_and_x,
+    unroll,
+    test_flags,
+    on_device,
+    fn_tree,
+    frontend,
+):
+    def _test_f(carry, x):
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:229]:
```suggestion
    def _test_scan_f(carry, x):
```

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 11]
```diff
@@ -203,3 +204,53 @@ def _test_body_fn(x):
...
+    fn_tree="jax.lax.scan",
+    dtype_and_x=helpers.dtype_and_values(
+        available_dtypes=helpers.get_dtypes("numeric"),
+        min_value=-1000,
+        max_value=1000,
+        min_num_dims=1,
+        min_dim_size=1,
+    ),
+    unroll=st.just(1),    
+)
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:219]:
```suggestion
    unroll=st.sampled_from([1, 2, 3]),
    reverse=st.booleans(),
)
```

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 12]
```diff
@@ -203,3 +204,53 @@ def _test_body_fn(x):
...
+def test_jax_lax_scan(
+    *,
+    dtype_and_x,
+    unroll,
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:223]:
```suggestion
    unroll,
    reverse,
```

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 13]
```diff
@@ -203,3 +204,53 @@ def _test_body_fn(x):
...
+
+    input_dtype, x = dtype_and_x
+    length = len(x)
+    expected_carry = 0
+    expected_ys = []
+    carry = expected_carry
+    for _ in range(unroll):
+        for elem in x:
+            carry, y = _test_f(carry, elem)
+            expected_ys.append(y)
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:242]:
```suggestion
    length = len(x[0])
```

File: ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py [hunk 14]
```diff
@@ -203,3 +204,53 @@ def _test_body_fn(x):
...
+        input_dtypes=input_dtype,
+        test_flags=test_flags,
+        frontend=frontend,
+        fn_tree=fn_tree,
+        on_device=on_device,
+        f=_test_f,
+        init=expected_carry,
+        xs=x,
+        length=length,
+        reverse=False,
```
Comment (Sam-Armstrong) [inline ivy_tests/test_ivy/test_frontends/test_jax/test_jax_lax_control_flow_operators.py:254]:
```suggestion
        f=_test_scan_f,
        init=x[0],
        xs=x[0],
        length=length,
        reverse=reverse,
```

**Review by Sam-Armstrong — CHANGES_REQUESTED** [review]
@TimothyElems I've left suggested change here which get a sample script working (will send on discord) and improve the tests so they pass when reverse=False and unroll=1. However, we need the frontend function to capture the full functionality of jax.lax.scan, and it currently isn't working correctly for reverse=True/unroll != 1. Please can you look into this and try to get it corrected? Thanks 🙂

**Review by CatB1t — CHANGES_REQUESTED** [review]
Hello @TimothyElems, could you please revert those changes to `dtype_helpers` module? the PR should be be only adding a new function to frontend API and not modify the testing pipeline.

**Comment by Sam-Armstrong** [comment]
@TimothyElems have you had chance to look at fixing the behaviour of 'unroll' and 'reverse' yet?

**Comment by ivy-seed** [comment]
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

**Comment by ivy-seed** [comment]
This PR has been closed because it has been marked as stale for more than 7 days with no activity.
